### ISSUE TITLE
Parsing negative intervals fails

### ISSUE THREAD

**Issue by manticore-projects** [duckdb/duckdb#11237]
### What happens?

Parsing negative intervals fails (unless quoted):
```sql
-- fails
SELECT DATE_ADD(DATE '2008-12-25', INTERVAL -5 DAY) AS five_days_ago
;
```

### To Reproduce

```sql
-- works
SELECT DATE_ADD(DATE '2008-12-25', INTERVAL 5 DAY) AS five_days_later
;

-- works
SELECT DATE_ADD(DATE '2008-12-25', INTERVAL '5 DAY') AS five_days_later
;

-- fails
SELECT DATE_ADD(DATE '2008-12-25', INTERVAL -5 DAY) AS five_days_ago
;

-- works
SELECT DATE_ADD(DATE '2008-12-25', INTERVAL '-5 DAY') AS five_days_ago
;
```

### OS:

Linux

### DuckDB Version:

0.10.0

### DuckDB Client:

Java

### Full Name:

Andreas Reichel

### Affiliation:

manticore-projects.com

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a release build (and could not test with a nightly build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by soerenwolfers** [comment]
You need to put parentheses around the number (same applies if you want to use a column in place of the number). Without having thought much about it, I do agree with you that it'd be nicer if it worked without the parentheses.
```
-- works
SELECT DATE_ADD(DATE '2008-12-25', INTERVAL (-5) DAY) AS five_days_ago
-- fails
SELECT DATE_ADD(DATE '2008-12-25', INTERVAL delta DAY) AS five_days_ago FROM (SELECT 5 AS delta)
-- works
SELECT DATE_ADD(DATE '2008-12-25', INTERVAL (delta) DAY) AS five_days_ago FROM (SELECT 5 AS delta)
-- works
SELECT DATE_ADD(DATE '2008-12-25', INTERVAL (delta) DAY) AS five_days_ago FROM (SELECT -5 AS delta)
```

**Comment by manticore-projects** [comment]
Thank you for prompt response, I appreciate.
Since I am working on `JSQLParser` I get somehow behind the requirement to put complex expressions into extra brackets.

However, parser wise the negative long should not be different from a positive long.
And also a column name should be a simple expression and not depend on extra brackets.

It is just extremely confusing especially as the Website https://duckdb.org/docs/sql/functions/date does not say anything about it (as far as I could see).

**Comment by manticore-projects** [comment]
I am sorry, I do not understand the bracket requirement at all:
```
INTERVAL := <INTERVAL_TOKEN> Expression <PART_TOKEN>
```

With `Expression` anything like `Long`, `String`, `Select`, `Column` or `XORExpression` etc.
I am actually more puzzled that `INTERVAL '5 days'` does work, since the `PART_TOKEN` was not present.

**Comment by Mytherin** [comment]
Unfortunately our current parser generator (yacc/bison) restricts us here, which is why the brackets are required. Postgres has more strict requirements and only allows string literals to be used with intervals:

```sql
postgres=# select interval '5' day;
 interval 
----------
 5 days
(1 row)

postgres=# select interval 5 day;
-- ERROR:  syntax error at or near "5"
-- LINE 1: select interval 5 day;
--                         ^
postgres=# select interval my_col day;
-- ERROR:  syntax error at or near "day"
-- LINE 1: select interval my_col day;
--                                ^
postgres=# select interval (my_col) day;
-- ERROR:  syntax error at or near "my_col"
-- LINE 1: select interval (my_col) day;
--                          ^

```

We have expanded the syntax to allow for arbitrary expressions, but they must be within brackets. I have tried and I do not believe this restriction can be alleviated without moving to a different parser generator which would entail rewriting the entire parser. We can improve the documentation around this.

**Comment by manticore-projects** [comment]
Thank you, I deeply appreciate your feedback and trust me, I feel you on that parser stuff (and had my own share of that).

Right now I am going through all DuckDB's date/time functions and I found that the quoting of the `DATE parts` and `INTERVAL` is confusing, since sometimes quotes are needed and sometimes not -- and I did not figure any pattern out.

Examples:

```sql
-- Why MUST DAY be quoted here?! (DAY is a reserved keyword and there should not be any confusion with a column or function)
SELECT DATE_DIFF('DAY', DATE '1970-01-01', DATE '2008-12-25') AS days_from_epoch
;
```

```sql
-- work?
SELECT DATE_ADD(DATE '1970-01-01', INTERVAL 1 DAY) AS date_from_epoch
;

-- works too!
SELECT DATE_ADD(DATE '1970-01-01', INTERVAL '1' DAY) AS date_from_epoch
;

-- And now its not?
SELECT DATE_ADD(DATE '1970-01-01', INTERVAL '-1 DAY') AS date_from_epoch
;

-- works too?
SELECT DATE_ADD(DATE '1970-01-01', INTERVAL '-1' DAY) AS date_from_epoch
;
```

My suggestion to clear this up was to recommend/ask for "always quoting" and only optionally/silently accept the cases, where its not needed.

**Comment by Mytherin** [comment]
`DAY` is not a reserved keyword:

```sql
D select date_diff(day, DATE '1970-01-01', DATE '2008-12-25') AS result from (values ('year')) t(day);
┌────────┐
│ result │
│ int64  │
├────────┤
│     38 │
└────────┘

```

**Comment by manticore-projects** [comment]
For my particular example:

```sql
SELECT Date_Diff( day, DATE '1970-01-01', DATE '2008-12-25' ) AS days_from_epoch
;
```

```text
java.sql.SQLException: Binder Error: Referenced column "DAY" not found in FROM clause!
LINE 2: SELECT DATE_DIFF(DAY, DATE '1970-01-01', DATE '2008-12-25') AS days_from_epoch...
                         ^
```

![image](https://github.com/duckdb/duckdb/assets/18080123/79baac63-ce10-492e-b06a-e84a44730093)

**Comment by soerenwolfers** [comment]
I think you're confusing two things here. 

`DATE_DIFF` is a regular function, whose first parameter is of string type. You can pass it a string literal as first argument, or you can pass it a column expression (a special case of which is just a column name). String literals require quotes, column names do not; but what arrives in the function in both cases is just the string content.

`INTERVAL X` is a INTERVAL constructor that can take many forms. If you only ever need to invoke this constructor with string literals you can do so. However, the documentation can't say that `INTERVAL '-123 DAY'` is how you should always write your interval constructors, and the syntax can't be limited to that, because the constructor is more powerful and allows things like `INTERVAL <COLUMN EXPRESSIONOFSTRINGTYPE>` (of which a string literal is a special case) or `INTERVAL <COLUMNEXPRESSIONOFINTEGERTYPE> <PARTKEYWORDINTHISCONTEXT>` (an integer literal for the middle expression is a special case). Note that this makes duckdb's interval constructor strictly more powerful, and what Mytherin pointed out is that at that added level of syntactic complexity, it becomes hard to distinguish the different constructs reliably, especially with third party parsers, which is why, for example, the last syntax I mentioned occasionally requires parentheses.

**Comment by manticore-projects** [comment]
Good Morning @soerenwolfers  and Team,

again thank you for your explanation and time. Technically everything is working fine and I only share my sentiment coming from other RDBMS.

Yes, you are correct on the "first parameter is of string type" although:

- the documentation says `date_diff(part, startdate, enddate)` and `part` being a technical term for unquoted DAY, WEEK, MONTH, YEAR etc. on most of the other RDBMS
- but only on the next page "Date Part Functions" the String requirement is "The fields are specified as **strings** that contain the part name of the field."
- at the same time `EXTRACT` is defined as `extract(part from date)` and here unquoted part works. (though I understand that `EXTRACT` liklely is parsed as an special expression and not as a regular function).


```sql
-- documentation says: extract(part from date)
-- works
SELECT EXTRACT(DAY FROM DATE '2013-12-25') AS the_day
;

-- documentation says: date_diff(part, startdate, enddate)
-- fails
SELECT DATE_DIFF(DAY, DATE '2008-12-25', DATE '2010-07-07') AS days_diff;
```

In summary we can close this case since I understand all the corner cases now.
But maybe we can document and illustrate such corner cases better. I am willing to help on this when needed.
